### ISSUE TITLE
Getting 500 error when using Guzzle but not when using cURL

### ISSUE THREAD

**Issue by allejo** [guzzle/guzzle#1841]
So I've been struggling with this for a bit. I can get GET requests to work no problem with Guzzle but I keep getting 500 errors when I try doing either POST or PUT requests. When configuring cURL manually, it works as intended but not when I try doing the same PUT request in guzzle. What am I doing wrong or where can I look to debug the issue?

```php
$url = 'http://domain.com/api';
$postString = 'value=my+New+Value';

// cURL

$curl = curl_init();

curl_setopt_array($curl, [
    CURLOPT_URL            => $url,
    CURLOPT_RETURNTRANSFER => true,
    CURLOPT_POST           => true,
    CURLOPT_POSTFIELDS     => $postString,
    CURLOPT_CUSTOMREQUEST  => 'PUT'
]);

$curlResult = curl_exec($curl);
print_r(json_decode($curlResult, true));

// Guzzle

$client = new Client();
$guzzleResult = $client->put($url, [
    'body' => $postString
]);

print_r(json_decode($guzzleResult->getBody(), true));
```

Edit: Currently using PHP 7.1.5

**Comment by sebastiaanluca** [comment]
I might have a related issue. Upgraded my PHP 7.1 installation today in production and all sites that use Guzzle started throwing 504 and 502 nginx errors :/

**Comment by sagikazarmark** [comment]
It's not completely clear to me: do you receive 500 errors from the server you send requests to with Guzzle or Guzzle causes your script to return with an 500 error.

In the first case it could be that the request is malformed (but even then the remote server is supposed to return 4xx errors). You should take a look at the remote server's log for the reason.

If it's the latter you should take a look at your logs to see why the request ends up in a 500. It's most likely an unhandled exception or error.

**Comment by sebastiaanluca** [comment]
Can confirm my issue is unrelated. Spent 2 days debugging and researching it to come to the conclusion all connections on the same server to the server's Digital Ocean floating IP failed, resulting in time-outs and other weird errors.

Turns out there was a sudden hiccup with the floating IP. Unassigning and reassigning it to the server did the trick.

**Comment by allejo** [comment]
Sorry, I'm receiving the 500 error from the URL I'm calling from guzzle. Unfortunately, it's not a URL I control so I can't view the logs. I'll see if I can reproduce the issue locally but if I can't, any other advice as to what I can do?

**Comment by sagikazarmark** [comment]
If Guzzle sends a malformed request the remote server should return a 4xx status code and in an optimal case an error message too. I would contact the maintainers of this service and ask for their help, notify them about this case, they might want to fix this issue (at least the status code) on their side.

If some kind of documentation to this service you could check the warning signs for things like "Content-Type header is mandatory", etc.

Last but not least you could dump the sent request in debug mode and send it to the service maintainers, see if they can spot anything extraordinary.

Unfortunately this is all I have and I know it's not much. It's hard to tell based on the provided information whether it's a bug in Guzzle or just a gone side effect that your code relied on.

**Comment by Skizu** [comment]
I'd like to bump this, I had this issue with a new production server returning unexpected 400 responses compared to the development and test environment working as expected; simply installing `apt install php7.0-curl` fixed it.

It was a brand new Ubuntu 16.04 LTS install with php installed via `ppa:ondrej/php`, during debugging I noticed that the headers were different. Both were sending a multi-part form with chucked data, however without `php7.0-curl` it was sending a `Connection: close` header rather than the `Expect: 100-Continue`; both requests of which had `Transfer-Encoding: chunked`.

**Comment by sagikazarmark** [comment]
It's not the same than: 400 is not the same as 500. 400 is returned when a request cannot be processed by the server because it is not formally valid. The server should also return some information or should have some log, otherwise it's quite hard to debug.

**Comment by areoid** [comment]
Maybe you should try this

```
try {
    $client = new Client();
    $guzzleResult = $client->put($url, [
        'body' => $postString
    ]);
} catch (\GuzzleHttp\Exception\RequestException $e) {
    $guzzleResult = $e->getResponse();
}

var_export($guzzleResult->getStatusCode());
var_export($guzzleResult->getBody());
```

Guzzle need cactching if the response code not 200

**Comment by iSWORD** [comment]
In my case it was because I had passed an empty array in the request's ` $options['json'] `
I couldn't reproduce the 500 on the serer using Postman or cURL even when passing the ` Content-Type: application/json ` request header.

Anyway, removing the json key from the request's options solved the problem.

I spent like 30 mins trying to figure out what's wrong because this behavior is very inconsistent. For all other request I'm making, passing ` $options['json'] = [] ` didn't cause any issues. It could be a server issue tho, I don't control the server.
